File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int result = Integer.MIN_VALUE;
        int n = heights.length ;
        int pse,nse ,ind;
        for(int i=0;i<n;i++){

            while(!stack.empty() && heights[i] <= heights[stack.peek()]){
                ind = stack.pop();
                nse = i ;
                pse = stack.isEmpty()? -1 : stack.peek();
                result = Math.max(result,(nse-pse-1)* heights[ind]);



            }
            stack.push(i);
        }
        while(!stack.isEmpty()){
            ind = stack.pop();
            pse = stack.isEmpty()? -1 : stack.peek();
            nse = n ;
            result = Math.max(result,(nse-pse-1)* heights[ind]);
        }
        return result;
        
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution(object):
        










    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        n =len(heights)
        pse = [-1] * n
        nse = [n] * n 
        stack = []
        for i in range(len(heights)) :
            while stack and heights[i] <= heights[stack[-1]] :
                nse[stack[-1]] = i 
                stack.pop()

            if stack :
                pse[i] = stack[-1]
            
            stack.append(i)





        result = 0
        for i in range(len(heights)) :
            result =max(result , ((nse[i] - pse[i]) -1) * heights[i])
        return result
        